config/cass-operator/cluster-scoped/kustomization.yaml

```diff
@@ -3,7 +3,38 @@ kind: Kustomization
 
 resources:
 - github.com/k8ssandra/cass-operator/config/deployments/cluster?ref=master
-
+patches:
```

Comment:
I share all of your concerns and this was definitely tricky for me to figure out, but it seems that "this is the way" 🤷 
Happy to start a discussion on how this could be improved.

Reply:
I'm not sure I understand the concerns. The overall approach is to extend the ComponentConfig as describe in the kubebuilder book [here](https://book.kubebuilder.io/component-config-tutorial/tutorial.html). 

Why would users need to change the kustomization file? ImageConfig declares defaults. If I want to change those defaults I can simply modify the cass-operator-manager-config ConfigMap. I would really prefer to see something like this over what we have in k8ssandra-operator. If I want images to always be pulled from a private registry, I only need to modify the ConfigMap. In k8ssandra-operator we either have to supply the image coordinates with each K8ssandraCluster (not DRY), or we have to recompile the operator to change the defaults defined in Go code. On top of that, default image coordinates are not centralized in one place in k8ssandra-operator.

Lastly, I'm not convinced of the need for a CRD; however, we really need docs to explain the structure and fields.

Reply:
I think we all agree we have a problem of insufficient documentation around operator configuration.

My main concern here is that the actual requirement is to change 2 lines in image_config.yaml:

```
defaults.dse.repository = "datastax/dse-mgmtapi-6_8"
defaults.dse.suffix     = ""
```

Unfortunately, that is not possible with kustomize (or is it?). We are then forced to re-declare the entire contents of image_config.yaml here. This is imho a fragile design that will force us to keep this file in sync with cass-operator every time the original file is updated with new defaults.

Reply:
The structure of the imageConfig is defined here: https://github.com/k8ssandra/cass-operator/blob/master/apis/config/v1beta1/imageconfig_types.go

Same as every CRD. We simply don't generate the Kubernetes yaml presentation.

In our Helm charts, we allow modification of single values in the imageConfig, since Helm renders the file. If you have to add all the fields in a patch, then that's more a limitation of Kustomize (and why it's not necessarily a good tool to deploy anything that requires kustomization (tm)). The issue is that Kustomize does not allow modifying files before the generator reads the input. 

That said, this approach seems complex what's done here. Wouldn't it be easier to do it like we patch the other field in this same ConfigMap when webhook is enabled?

https://github.com/k8ssandra/cass-operator/tree/master/config/components/webhook

It uses a file as an input to modify the OperatorConfig.